### ISSUE TITLE
RequestDelegateGenerator generate incorrect code if use with fully specify typed Results

### ISSUE THREAD

**Issue by NotAsea** [dotnet/aspnetcore#52989]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

the code bellow work if i dont enable `EnableRequestDelegateGenerator`
```cs
//Program.cs
using TestRepo.Routes;
using TestRepo.Setup;

var builder = WebApplication.CreateBuilder(args);

builder.RegisterService(); // typical register service builder.Service.Add{stuf} i put it in another file to keep thing clean

var app = builder.Build();

await app.StartupAction();  // this check DB and do seed data

app.MapGroup("/person").WithTags("Person").HandlePersonRoute(); // the main route define bellow

app.Run();
```
```cs
// PersonRoute.cs
public static class PersonRoute
{
    /// <summary>
    /// this consumes the <see cref="RouteGroupBuilder"/> and handle all logic and child route. <br />
    /// This method must be and mean to be called last of <c>Map{Verb}</c> chain, as it return <see cref="Void"/> 
    /// </summary>
    /// <param name="route"></param>
    public static void HandlePersonRoute(this RouteGroupBuilder route)
    {
         route.MapPost("/add", CreatePerson);
        //  more route handler like above
    }

    private static async Task<Results<Ok<int>, BadRequest<string>>> CreatePerson(
        ILogger logger,
        IRepository repository,
        Person person
    )
    {
       // do validate

      // -----snif--------

        await using var transaction = await repository.BeginTransactionAsync();
        try
        {
            await repository.AddAsync(person);
            await repository.SaveChangesAsync();
            await transaction.CommitAsync();
            return TypedResults.Ok(person.Id);
        }
        catch (Exception ex)
        {
            var reason = ex.GetBaseException().Message;
            await transaction.RollbackAsync();
            logger.WriteToDatabaseFail(nameof(Person), reason);
            return TypedResults.BadRequest(reason);
        }
    }
// --- snif-----
}
```

### Expected Behavior

the app build and run perfectly fine if using default template, but fail to compile if enable either `EnableRequestDelegateGenerator` or `PublishTrimmed` or `PublishAOT`, note that i put EFCore in another project and reference to it
below is compiler complain
![image](https://github.com/dotnet/aspnetcore/assets/88715515/8187c49b-7af9-4c6f-9569-99c89410d43a)
and in source gen file
![image](https://github.com/dotnet/aspnetcore/assets/88715515/331d1431-c8af-43e9-af39-c73d26e7a451)
I dont know how `false` can be generate as `False` but i suspect something to do with `TypedResults`

### Steps To Reproduce

you can find my repo here https://github.com/NotAsea/TestRepo.git
 - create with Web API template in Visual Studio
 - dont choose enable native AOT
 - dont choose use Controller 

### .NET Version

```bash
.NET SDK:
 Version:           8.0.100
 Commit:            57efcf1350
 Workload version:  8.0.100-manifests.8d38d0cc

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.22621
 OS Platform: Windows
 RID:         win-x64
 Base Path:   C:\Program Files\dotnet\sdk\8.0.100\

.NET workloads installed:
 Workload version: 8.0.100-manifests.8d38d0cc
There are no installed workloads to display.

Host:
  Version:      8.0.0
  Architecture: x64
  Commit:       5535e31a71

.NET SDKs installed:
  8.0.100 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 6.0.25 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 7.0.14 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 8.0.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 6.0.25 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.14 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 8.0.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 6.0.25 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 7.0.14 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 8.0.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

Other architectures found:
  x86   [C:\Program Files (x86)\dotnet]
    registered at [HKLM\SOFTWARE\dotnet\Setup\InstalledVersions\x86\InstallLocation]

Environment variables:
  Not set

global.json file:
  Not found
```

**Comment by NotAsea** [comment]
for more context `RequestDelegateGenerator` do work if i don't split handler to another function, and **DON'T** specify type to Delegate like so 
```cs
// PersonRoute.cs
public static class PersonRoute
{
    /// <summary>
    /// this consumes the <see cref="RouteGroupBuilder"/> and handle all logic and child route. <br />
    /// This method must be and mean to be called last of <c>Map{Verb}</c> chain, as it return <see cref="Void"/> 
    /// </summary>
    /// <param name="route"></param>
    public static void HandlePersonRoute(this RouteGroupBuilder route, ILogger logger)
    {
         CreatePerson(route, logger);
         // route.MapPost("/add", CreatePerson);
        //  more route handler like above
    }

    private static void CreatePerson(IEndpointRouteBuilder route, ILogger logger)
    {
        route.MapPost("/add", async (
            ILogger logger,
            IRepository repository,
            Person person
        ) => {
          // just verson 1 but move to body of Delegate
        });
    }
// --- snif-----
}
```
Either of these, i see no official document mentioning how you can split route in Minimal API **AND DO IT WITHOUT BREAKING**

**Comment by joegoldman2** [comment]
Duplicate of #51183.

This issue has already been fixed in #51193 for .NET 9 preview 1. The fix has still not been backported to 8.0.x. Probably a small oversight as it was planned.

cc @captainsafia

**Comment by captainsafia** [comment]
> This issue has already been fixed in https://github.com/dotnet/aspnetcore/pull/51193 for .NET 9 preview 1. The fix has still not been backported to 8.0.x. Probably a small oversight as it was planned.

Thanks for the nudge here, @joegoldman2! It does look like this missed going through the backport process. :/ I've initiated that with the hope of it going into the next release.

> for more context RequestDelegateGenerator do work if i don't split handler to another function, and DON'T specify type to Delegate like so

It's not clear to me what doesn't work refers to here? Do you get warnings when compiling with AoT from the linker analyzer or is it just the CS0103 errors mentioned above?

**Comment by captainsafia** [comment]
I just remembered why I hadn't backported it yet. I wanted to give the changeset some "bake-time" in the 9.0 branch before backporting since I took a slightly different direction on how default values are generated than other source generators have.

@NotAsea Is there a chance you can try the 9.0-preview1 nightly and confirm if that addresses the issues that you're seeing?

You can find the nightly installer at https://github.com/dotnet/installer.
